srml/babe/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+use inherents::{RuntimeString, InherentIdentifier, InherentData, ProvideInherent, MakeFatalError};
+#[cfg(feature = "std")]
+use inherents::{InherentDataProviders, ProvideInherentData};
+
+/// The babe inherent identifier.
+pub const INHERENT_IDENTIFIER: InherentIdentifier = *b"babeslot";
+
+/// The type of the babe inherent.
+pub type InherentType = u64;
+
+/// Auxiliary trait to extract babe inherent data.
+pub trait BabeInherentData {
+	/// Get babe inherent data.
+	fn babe_inherent_data(&self) -> result::Result<InherentType, RuntimeString>;
+	/// Replace babe inherent data.
+	fn babe_replace_inherent_data(&mut self, new: InherentType);
+}
+
+impl BabeInherentData for InherentData {
+	fn babe_inherent_data(&self) -> result::Result<InherentType, RuntimeString> {
+		self.get_data(&INHERENT_IDENTIFIER)
+			.and_then(|r| r.ok_or_else(|| "Babe inherent data not found".into()))
+	}
+
+	fn babe_replace_inherent_data(&mut self, new: InherentType) {
+		self.replace_data(INHERENT_IDENTIFIER, &new);
+	}
+}
+
+/// Provides the slot duration inherent data for `Babe`.
+#[cfg(feature = "std")]
+pub struct InherentDataProvider {
+	slot_duration: u64,
+}
+
+#[cfg(feature = "std")]
+impl InherentDataProvider {
+	pub fn new(slot_duration: u64) -> Self {
+		Self {
+			slot_duration
+		}
+	}
+}
+
+#[cfg(feature = "std")]
+impl ProvideInherentData for InherentDataProvider {
+	fn on_register(
+		&self,
+		providers: &InherentDataProviders,
+	) -> result::Result<(), RuntimeString> {
+		if !providers.has_provider(&timestamp::INHERENT_IDENTIFIER) {
+			// Add the timestamp inherent data provider, as we require it.
+			providers.register_provider(timestamp::InherentDataProvider)
+		} else {
+			Ok(())
+		}
+	}
+
+	fn inherent_identifier(&self) -> &'static inherents::InherentIdentifier {
+		&INHERENT_IDENTIFIER
+	}
+
+	fn provide_inherent_data(
+		&self,
+		inherent_data: &mut InherentData,
+	) -> result::Result<(), RuntimeString> {
+		let timestamp = inherent_data.timestamp_inherent_data()?;
+		let slot_num = timestamp / self.slot_duration;
+		inherent_data.put_data(INHERENT_IDENTIFIER, &slot_num)
+	}
+
+	fn error_to_string(&self, error: &[u8]) -> Option<String> {
+		RuntimeString::decode(&mut &error[..]).map(Into::into)
+	}
+}
+
+/// Something which can handle Babe consensus reports.
+pub trait HandleReport {
+	fn handle_report(report: BabeReport);
+}
+
+impl HandleReport for () {
+	fn handle_report(_report: BabeReport) { }
+}
+
+pub trait Trait: timestamp::Trait {
+	/// The logic for handling reports.
+	type HandleReport: HandleReport;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Babe {
+		// The last timestamp.
+		LastTimestamp get(last) build(|_| T::Moment::sa(0)): T::Moment;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin { }
+}
+
+/// A report of skipped authorities in babe.
+#[derive(Clone, Encode, Decode, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct BabeReport {
```

Comment:
Unfortunately, BABE doesn't work like this (since you can't go from slot -> author like you can in Aura). All the punishment machinery here should be removed for now.